components/builder-minio/habitat/config_install/minio-migration.sh

```diff
@@ -0,0 +1,116 @@
+#!/bin/bash
+
+function _hab_exec() {
+  pkg=$1
+  shift
+  hab pkg exec "$pkg" -- "$@"
+}
+
+function _aws() {
+  _hab_exec 'core/aws-cli' aws --endpoint-url "{{cfg.env.MINIO_ENDPOINT}}" "$@"
+}
+
+function _curl() {
+  _hab_exec 'core/curl' curl "$@"
+}
+
+function _jq() {
+  _hab_exec 'core/jq-static' jq "$@"
+}
+
+function download_bucket_objects() {
+  echo "DOWNLOADING objects from the MinIO that we are migrating from"
+  if ! _aws s3 sync "$s3_url" "$WAYPOINT"; then
+    echo "ERROR: download of objects FAILED"
+    exit 8
+  fi
+  return 0
+}
+
+function _ensure_bucket_exists() {
+  if ! _aws s3 ls "$MINIO_BUCKET" &>/dev/null; then
+    _aws s3 mb "$s3_url"
+  fi
+}
+
+function upload_bucket_objects() {
+  _ensure_bucket_exists
+  echo "UPLOADING objects to the MinIO that we are migrating to"
+  if ! _aws s3 sync "$WAYPOINT" "$s3_url"; then
+    echo "ERROR: upload of objects FAILED"
+    exit 9
+  fi
+  return 0
+}
+
+function is_migration_from_removed_fs_backend_needed() {
+  echo "CHECKING if migration from removed fs backend is needed"
+  if [[ -f /hab/svc/builder-minio/data/.minio.sys/format.json ]]; then
+    format_value=$(_jq -r '.format' /hab/svc/builder-minio/data/.minio.sys/format.json)
+    if [[ "${format_value}" == 'fs' ]]; then
+      return 0
+    fi
+  fi
+  return 1
+}
+
+function minio_health_live_check() {
+  for ((n = 0; n < 20; n++)); do
+    curl_http_code_args=(-fs -o /dev/null -w "%{http_code}" --retry 4 --retry-delay 1)
+    code=$(_curl "${curl_http_code_args[@]}" "$MINIO_ENDPOINT/minio/health/live")
+    if [[ $code == 200 ]]; then
+      return 0
+    else
+      sleep .5
+    fi
+  done
+  return 1
+}
+
+function minio_stop() {
+  for ((n = 0; n < 20; n++)); do
+    if pgrep minio >/dev/null; then
+      pkill minio >/dev/null
+      sleep 1
+    else
+      return 0
+    fi
+  done
+}
+
+function config_environment_for_migration() {
+  if [[ ! $1 =~ ^[0-9]+$ ]]; then
+    echo "ERROR: Invalid timestamp"
+  fi
+  export MINIO_ENDPOINT="{{cfg.env.MINIO_ENDPOINT}}"
+  export AWS_ACCESS_KEY_ID="{{cfg.env.MINIO_ACCESS_KEY}}"
+  export AWS_SECRET_ACCESS_KEY="{{cfg.env.MINIO_SECRET_KEY}}"
+  export s3_url="s3://{{cfg.env.MINIO_BUCKET}}"
+  export MINIO_BUCKET="{{cfg.env.MINIO_BUCKET}}"
```

Comment:
would the two above be different from `{{cfg.bucket_name}}`?

Reply:
In our hooks we have that habitat-builder-artifact-store.default but in bldr.env.sample and the scripting of on-prem-builder we have habitat-builder-artifact-store.local. I think the user.toml files probably explain why this isn't biting us today.

Reply:
I didn't notice that the `bucket_name` in the default.toml is slightly different than what is written to the user.toml. However, the hooks just use what is in `bucket_name` so I'd think you would want to use whatever value is stored in the `bucket_name` configuration setting because that is where the artifacts will be.

Reply:
I'm not sure about this one in particular. In my on prem builder the bucket name is habitat-builder-artifact-store.local so I'm really feeling like using the info from the bldr.env is the way to go, at least for here.  I lean that way for all of it but when it comes to the bucket name I think we have to use what's coming from bldr.env and while I feel the same about the minio endpoint stuff I would want to review a bit more yet.

Reply:
I guess what I'm trying to say is that if a user has used a bldr.env to set up the user.toml, then `{{cfg.bucket_name}}` will equal `habitat-builder-artifact-store.local` and thus there is no need to introduce a separate setting.

Reply:
OK, you seem certain and I'm confidence level varies based on which exact part of all this we're talking about.  I'll just try to make everything here way you've asked for.

Reply:
Done.